src/storage/lock_manager/deadlock.rs

```diff
@@ -22,91 +22,158 @@ use std::fmt::{self, Display, Formatter};
+        pushed.insert(wait_for_ts);
+        while let Some(wait_for_ts) = stack.pop() {
+            if let Some(wait_for) = self.wait_for_map.get_mut(&wait_for_ts) {
+                // Remove expired edges.
+                wait_for.retain(|_, locks| !locks.is_expired(now, ttl));
+                if wait_for.is_empty() {
+                    self.wait_for_map.remove(&wait_for_ts);
+                } else {
+                    for (lock_ts, locks) in wait_for {
+                        if *lock_ts == txn_ts {
+                            return Some(locks.hashes[0]);
+                        }
+                        if !pushed.contains(lock_ts) {
+                            stack.push(*lock_ts);
+                            pushed.insert(*lock_ts);
+                        }
+                    }
                 }
             }
         }
         None
     }
 
-    /// Returns true and adds to the detect table if txn_ts is waiting for lock_ts.
+    /// Returns true and adds to the detect table if `txn_ts` is waiting for `lock_ts`.
     fn register_if_existed(&mut self, txn_ts: u64, lock_ts: u64, lock_hash: u64) -> bool {
         if let Some(wait_for) = self.wait_for_map.get_mut(&txn_ts) {
-            if let Some(lock_hashes) = wait_for.get_mut(&lock_ts) {
-                if !lock_hashes.contains(&lock_hash) {
-                    lock_hashes.push(lock_hash);
-                }
+            if let Some(locks) = wait_for.get_mut(&lock_ts) {
+                locks.push(lock_hash, self.now);
                 return true;
             }
         }
         false
     }
 
-    /// Adds to the detect table.
+    /// Adds to the detect table. The edge from `txn_ts` to `lock_ts` must not exist.
     fn register(&mut self, txn_ts: u64, lock_ts: u64, lock_hash: u64) {
         let wait_for = self.wait_for_map.entry(txn_ts).or_default();
-        let lock_hashes = wait_for.entry(lock_ts).or_default();
-        if !lock_hashes.contains(&lock_hash) {
-            lock_hashes.push(lock_hash);
-        }
+        assert!(!wait_for.contains_key(&lock_ts));
+        let locks = Locks::new(lock_ts, lock_hash, self.now);
+        wait_for.insert(locks.ts, locks);
     }
 
     /// Removes the corresponding wait_for_entry.
     pub fn clean_up_wait_for(&mut self, txn_ts: u64, lock_ts: u64, lock_hash: u64) {
         if let Some(wait_for) = self.wait_for_map.get_mut(&txn_ts) {
-            if let Some(lock_hashes) = wait_for.get_mut(&lock_ts) {
-                let idx = lock_hashes.iter().position(|hash| *hash == lock_hash);
-                if let Some(idx) = idx {
-                    lock_hashes.remove(idx);
-                    if lock_hashes.is_empty() {
-                        wait_for.remove(&lock_ts);
-                        if wait_for.is_empty() {
-                            self.wait_for_map.remove(&txn_ts);
-                        }
+            if let Some(locks) = wait_for.get_mut(&lock_ts) {
```

Comment:
The code nests too deep. You can use `and_then` instead.

Reply:
I can't make it. 😥

Reply:
I appreciate if you can try it. @sticnarf

Reply:
I think it's uglier:
```rust
        if self
            .wait_for_map
            .get_mut(&txn_ts)
            .and_then(|wait_for| {
                if let Some(locks) = wait_for.get_mut(&lock_ts) {
                    if locks.remove(lock_hash) {
                        wait_for.remove(&lock_ts);
                    }
                }
                Some(wait_for.is_empty())
            })
            .unwrap_or(false)
        {
            self.wait_for_map.remove(&txn_ts);
        }
```

Reply:
```rust
        (|| {
            let wait_for = self.wait_for_map.get_mut(&txn_ts)?;
            let locks = wait_for.get_mut(&lock_ts)?;
            if locks.remove(lock_hash) {
                wait_for.remove(&lock_ts);
                if wait_for.is_empty() {
                    self.wait_for_map.remove(&txn_ts);
                }
            }
            Some(())
        })();
```

Here is a small trick which you can try. In fact I don't like it very much, though...

Reply:
It's not a big improvement. I will keep it for simplicity.